### PR TITLE
(maint) Make sure aptly isn't running before we try to update

### PR DESCRIPTION
**Pull request by underscorgan**
This adds a loop to make sure aptly isn't running before we try to lock
the database.  LevelDB doesn't remove/re-add its lockfile each time it
takes/releases the lock, rather it seems to be doing some filesystem
locking.

### CODE REVIEW

File: tasks/pe_ship.rake [hunk 1]
```diff
@@ -159,13 +159,19 @@ if Pkg::Config.build_pe
         incoming_dir or fail "Adding packages to apt repo requires an incoming directory"
         Pkg::Util::RakeUtils.invoke_task("pl:fetch")
 
-        stdout, stderr = Pkg::Util::Net.remote_ssh_cmd(Pkg::Config.apt_host, "aptly repo add -remove-files #{Pkg::Config::pe_version}-#{dist} #{incoming_dir}/#{dist}
+        cmd = <<-eos
+          while pgrep aptly > /dev/null; do
```
Comment (shrug) [inline tasks/pe_ship.rake]: We can use flock() semantics?

Comment (mckern) [inline tasks/pe_ship.rake]: ^^ What he said

Comment (underscorgan) [inline tasks/pe_ship.rake]: I'm going to dig more in the db implementation to try to figure out what they're actually calling to lock the files.

Comment (mckern) [inline tasks/pe_ship.rake]: I was working backwards from this thread: https://groups.google.com/forum/#!topic/aptly-discuss/Gsemphn9CEo

**Comment by underscorgan** [comment]
@shrug @mckern updated

**Comment by mckern** [comment]
kum4bxe

Does it blend? It looks like it should blend.

**Comment by underscorgan** [comment]
I'm not sure how to test this without merging it since it's in the promotion codepath.
`jenkins@saturn:~$ flock /opt/tools/aptly/db/LOCK -c /bin/true` executes successfully though.

**Comment by mckern** [comment]
@underscorgan suggestion:

On any system with Aptly and a pile of packages on it, kick off an Aptly `repo -add` in the background/another tab or session. Run a second Aptly `repo -add` with the `flock` code in a separate session. See if it blends or if it explodes in a shower of sparks and matchbox car parts?

**Comment by underscorgan** [comment]
@mckern good call. Spun up a vm, mirrored PC1 for wheezy, imported that into a local repo. That took a few minutes. In another window started `flock .aptly/db/LOCK -c "date"`, it returned after the publishing of the local repo finished.

File: tasks/pe_ship.rake [hunk 2]
```diff
@@ -159,13 +159,16 @@ if Pkg::Config.build_pe
         incoming_dir or fail "Adding packages to apt repo requires an incoming directory"
         Pkg::Util::RakeUtils.invoke_task("pl:fetch")
 
-        stdout, stderr = Pkg::Util::Net.remote_ssh_cmd(Pkg::Config.apt_host, "aptly repo add -remove-files #{Pkg::Config::pe_version}-#{dist} #{incoming_dir}/#{dist}
+        cmd = <<-eos
+          flock /opt/tools/aptly/db/LOCK -c /bin/true
```
Comment (shrug) [inline tasks/pe_ship.rake]: Do you think we should have a timeout here? I don't have an opinion on what might be sane... something long-ish?

Comment (underscorgan) [inline tasks/pe_ship.rake]: 15 minutes? I feel like that should be long enough, and probably better to let the job explode in failing to update aptly than hang?

File: tasks/pe_ship.rake [hunk 3]
```diff
@@ -159,13 +159,19 @@ if Pkg::Config.build_pe
         incoming_dir or fail "Adding packages to apt repo requires an incoming directory"
         Pkg::Util::RakeUtils.invoke_task("pl:fetch")
 
-        stdout, stderr = Pkg::Util::Net.remote_ssh_cmd(Pkg::Config.apt_host, "aptly repo add -remove-files #{Pkg::Config::pe_version}-#{dist} #{incoming_dir}/#{dist}
+        cmd = <<-eos
+          if ! flock /opt/tools/aptly/db/LOCK --wait 1200 --command /bin/true; then
+            2&> echo "Unable to acquire aptly lock, giving up"
+            exit 1
+          fi
```
Comment (underscorgan) [inline tasks/pe_ship.rake]: @shrug this look reasonable?

File: tasks/pe_ship.rake [hunk 4]
```diff
@@ -159,13 +159,19 @@ if Pkg::Config.build_pe
         incoming_dir or fail "Adding packages to apt repo requires an incoming directory"
         Pkg::Util::RakeUtils.invoke_task("pl:fetch")
 
-        stdout, stderr = Pkg::Util::Net.remote_ssh_cmd(Pkg::Config.apt_host, "aptly repo add -remove-files #{Pkg::Config::pe_version}-#{dist} #{incoming_dir}/#{dist}
+        cmd = <<-eos
+          if ! flock /opt/tools/aptly/db/LOCK --wait 1200 --command /bin/true; then
+            2&> echo "Unable to acquire aptly lock, giving up"
```
Comment (mckern) [inline tasks/pe_ship.rake]: This is not an "output to STDERR" format that I've seen before. Look at that. But it doesn't work.

```bash
mckern@flexo errors $ ls -1
echo_errors.sh
mckern@flexo errors $ bash ./echo_errors.sh
mckern@flexo errors $ bash -x ./echo_errors.sh
+ 2 'Unable to acquire aptly lock, giving up'
mckern@flexo errors $ echo
mckern@flexo errors $ ls -1
echo
echo_errors.sh
mckern@flexo errors $ cat echo
echo            echo_errors.sh
mckern@flexo errors $ cat echo
./echo_errors.sh: line 3: 2: command not found
mckern@flexo errors $
```

Comment (underscorgan) [inline tasks/pe_ship.rake]: I'm not real familiar with bash outputting to stderr, this is what I found on stackoverflow. Do you have one that works/you like?

Comment (mckern) [inline tasks/pe_ship.rake]: Here's an output redirector cheatsheat I cribbed from somewhere else:

| Stream         | Description                                            |
|----------------|--------------------------------------------------------|
| `cmd > file` | Redirect the standard output (stdout) of cmd to a file.|
| `cmd 2>&1` | Redirect stderr to stdout. `2` is the file descriptor for stderr.|
| `cmd 1>&2` | Redirect stdout to stderr.|
| `cmd >&2` | Same as `cmd 1>&2`. `1` can often be omitted when redirecting stderr.|
| `cmd 1> file` | Same as `cmd > file`. `1` is the default file descriptor (fd) for stdout.|
| `cmd 2> file` | Redirect the standard error (stderr) of cmd to a file. 2 is the default fd for stderr.|
| `cmd >> file` | Append stdout of cmd to a file.|
| `cmd 2>> file` | Append stderr of cmd to a file.|
| `cmd > file 2>&1` | Redirect both stdout and stderr of cmd to a file. This is not the same as cmd 2>&1 > file. Redirection order matters!|

Comment (mckern) [inline tasks/pe_ship.rake]: (FWIW, redirection at the front of the line will work if constructed like this:

```bash
mckern@flexo errors $ bash ./echo_errors.sh >/dev/null
Front of line redirection
mckern@flexo errors $ cat echo_errors.sh
#!/bin/bash

>&2 echo "Front of line redirection"

mckern@flexo errors $
```

Comment (underscorgan) [inline tasks/pe_ship.rake]: ah, cool. thank you. I think I got to over thinking this one.

**Comment by underscorgan** [comment]
@mckern updated!

**Pull request merged**